sql: correctly handle CREATE INDEX on REGIONAL BY ROW tables

otan opened the pull request:
sql: correctly handle CREATE INDEX on REGIONAL BY ROW tables

* Refactored the code that translates from TableDesc to tree.PartitionBy
  without needing a parsing layer, because it does not work for REGIONAL
  BY ROW tables as the clause is omitted entirely.
* Introduce a way of altering the zone configurations without shelling
  out to another SQL call. We should update our other code to do the
  same logic, which I've added a TODO for.
* Modify CREATE INDEX to set the partitioning on the newly created
  indexes.

Release note (sql change): CREATE INDEX on REGIONAL BY ROW tables will
now correctly include the implicit partitioning and inherit the correct
zone configurations.



Resolves #58337

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/59223)
<!-- Reviewable:end -->

ajstorm reviewed (approved):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/59223)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @otan)

___
*[pkg/ccl/logictestccl/testdata/logic_test/regional_by_row, line 207 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/59223#-MRe6lO01ht0frPbOHc0:-MRe6lO01ht0frPbOHc1:bty10ar) ([raw file](https://github.com/cockroachdb/cockroach/blob/52af1f2d2f6a995b9c0f53afc512c13d9c9c1f28/pkg/ccl/logictestccl/testdata/logic_test/regional_by_row#L207)):*
> ```
> 
> query T
> SELECT create_statement FROM [SHOW CREATE TABLE regional_by_row_table]
> ```

Obviously not for this review, but it occurred to me while looking at this test case that it would be good if we had a bunch of tests to ensure that all of the new syntax we're adding properly round trips.  Do you know of any tests that we have to validate round trippiness?  If not, I think we should open an issue for that.  I can open it, if you validate that it's something we need.
___
*[pkg/ccl/logictestccl/testdata/logic_test/regional_by_row, line 256 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/59223#-MRe7XXlAnfKv72dwvAi:-MRe7XXlAnfKv72dwvAj:b-2dievy) ([raw file](https://github.com/cockroachdb/cockroach/blob/52af1f2d2f6a995b9c0f53afc512c13d9c9c1f28/pkg/ccl/logictestccl/testdata/logic_test/regional_by_row#L256)):*
> ```
>   constraints = '{+region=us-east-1: 1}',
>   lease_preferences = '[[+region=us-east-1]]';
> ALTER PARTITION "ap-southeast-2" OF INDEX multi_region_test_db.public.regional_by_row_table@new_idx CONFIGURE ZONE USING
> ```

Nit: It's a bit strange to me that these show up at the end.  Is that a result of the fact that new_idx was created most recently?  Is there some way that we could sort these either by region, or index name (I think region might make it the most readable).
___
*[pkg/sql/create_index.go, line 553 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/59223#-MRe8XnuAhzuTqI62S-p:-MRe8XnuAhzuTqI62S-q:b-2388ng) ([raw file](https://github.com/cockroachdb/cockroach/blob/52af1f2d2f6a995b9c0f53afc512c13d9c9c1f28/pkg/sql/create_index.go#L553)):*
> ```Go
> // configureZoneConfigForNewIndexPartitioning configures the zone config for any new index
> // in a REGIONAL BY ROW table.
> // This *must* be done after the index ID has been allocated.
> ```

Could we add an assertion here?  What will happen if it's called before the index ID is allocated?  A panic in `addNewZoneConfigSubzonesForIndex`?
___
*[pkg/sql/partition.go, line 31 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/59223#-MReAG3yACFUqnN0-ih7:-MReAG3yACFUqnN0-ih8:b8m8zaa) ([raw file](https://github.com/cockroachdb/cockroach/blob/52af1f2d2f6a995b9c0f53afc512c13d9c9c1f28/pkg/sql/partition.go#L31)):*
> ```Go
> }
> 
> func partitionByFromTableDescImpl(
> ```

Nit: This function is mildly complex, and is fairly sparsely commented.  Might be nice to have a comment above it which describes what's going on inside.
___
*[pkg/sql/region_util_test.go, line 437 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/59223#-MReCWEy89rHg5EK_MeP:-MReCWEy89rHg5EK_MeQ:b-uj3qx) ([raw file](https://github.com/cockroachdb/cockroach/blob/52af1f2d2f6a995b9c0f53afc512c13d9c9c1f28/pkg/sql/region_util_test.go#L437)):*
> ```Go
> 		},
> 		{
> 			desc: "4-region global table with region survivability",
> ```

oh sure, this doesn't get caught by the linter, but flavoured is out of bounds!?!


<!-- Sent from Reviewable.io -->

otan reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/59223)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajstorm)

___
*[pkg/ccl/logictestccl/testdata/logic_test/regional_by_row, line 207 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/59223#-MRe6lO01ht0frPbOHc0:-MRq5UZK9O6Q2V8hZ0c5:brpoavd) ([raw file](https://github.com/cockroachdb/cockroach/blob/52af1f2d2f6a995b9c0f53afc512c13d9c9c1f28/pkg/ccl/logictestccl/testdata/logic_test/regional_by_row#L207)):*
<details><summary><i>Previously, ajstorm wrote…</i></summary><blockquote>

Obviously not for this review, but it occurred to me while looking at this test case that it would be good if we had a bunch of tests to ensure that all of the new syntax we're adding properly round trips.  Do you know of any tests that we have to validate round trippiness?  If not, I think we should open an issue for that.  I can open it, if you validate that it's something we need.
</blockquote></details>

nah, there's a TODO somewhere.
in theory if we end up going down that course we should probably also not show ALTER PARTITIONS 
___
*[pkg/ccl/logictestccl/testdata/logic_test/regional_by_row, line 256 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/59223#-MRe7XXlAnfKv72dwvAi:-MRq5fRZ0hZbvX_YBKJP:b-1sjs94) ([raw file](https://github.com/cockroachdb/cockroach/blob/52af1f2d2f6a995b9c0f53afc512c13d9c9c1f28/pkg/ccl/logictestccl/testdata/logic_test/regional_by_row#L256)):*
<details><summary><i>Previously, ajstorm wrote…</i></summary><blockquote>

Nit: It's a bit strange to me that these show up at the end.  Is that a result of the fact that new_idx was created most recently?  Is there some way that we could sort these either by region, or index name (I think region might make it the most readable).
</blockquote></details>

Yep, The way these are sorted are more internal, if it bothers us we can make an issue for it. but if the round trip thing above is an issue we probably should get rid of it.
___
*[pkg/sql/create_index.go, line 553 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/59223#-MRe8XnuAhzuTqI62S-p:-MRqAgh64yjfpOZB1W1C:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/52af1f2d2f6a995b9c0f53afc512c13d9c9c1f28/pkg/sql/create_index.go#L553)):*
<details><summary><i>Previously, ajstorm wrote…</i></summary><blockquote>

Could we add an assertion here?  What will happen if it's called before the index ID is allocated?  A panic in `addNewZoneConfigSubzonesForIndex`?
</blockquote></details>

Done.
___
*[pkg/sql/region_util_test.go, line 437 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/59223#-MReCWEy89rHg5EK_MeP:-MRq9EF362P_m-dcPNme:b-az7e3m) ([raw file](https://github.com/cockroachdb/cockroach/blob/52af1f2d2f6a995b9c0f53afc512c13d9c9c1f28/pkg/sql/region_util_test.go#L437)):*
<details><summary><i>Previously, ajstorm wrote…</i></summary><blockquote>

oh sure, this doesn't get caught by the linter, but flavoured is out of bounds!?!
</blockquote></details>

goddim :)


<!-- Sent from Reviewable.io -->

otan committed:
sql: correctly handle CREATE INDEX on REGIONAL BY ROW tables

* Refactored the code that translates from TableDesc to tree.PartitionBy
  without needing a parsing layer, because it does not work for REGIONAL
  BY ROW tables as the clause is omitted entirely.
* Introduce a way of altering the zone configurations without shelling
  out to another SQL call. We should update our other code to do the
  same logic, which I've added a TODO for.
* Modify CREATE INDEX to set the partitioning on the newly created
  indexes.

Release note (sql change): CREATE INDEX on REGIONAL BY ROW tables will
now correctly include the implicit partitioning and inherit the correct
zone configurations.

otan commented:
bors r=ajstorm

ajstorm reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/59223)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajstorm and @otan)

___
*[pkg/ccl/logictestccl/testdata/logic_test/regional_by_row, line 207 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/59223#-MRe6lO01ht0frPbOHc0:-MRtlHoDBsOY2y5kXVK8:bobskqp) ([raw file](https://github.com/cockroachdb/cockroach/blob/52af1f2d2f6a995b9c0f53afc512c13d9c9c1f28/pkg/ccl/logictestccl/testdata/logic_test/regional_by_row#L207)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

nah, there's a TODO somewhere.
in theory if we end up going down that course we should probably also not show ALTER PARTITIONS 
</blockquote></details>

I added #59372.  I'm starting to worry that we've got a bunch of TODOs which require work which we can't see from our GitHub board.  
___
*[pkg/ccl/logictestccl/testdata/logic_test/regional_by_row, line 256 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/59223#-MRe7XXlAnfKv72dwvAi:-MRtlv0e83YV-pJPWQYA:brz66yh) ([raw file](https://github.com/cockroachdb/cockroach/blob/52af1f2d2f6a995b9c0f53afc512c13d9c9c1f28/pkg/ccl/logictestccl/testdata/logic_test/regional_by_row#L256)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

Yep, The way these are sorted are more internal, if it bothers us we can make an issue for it. but if the round trip thing above is an issue we probably should get rid of it.
</blockquote></details>

Added to #59372 


<!-- Sent from Reviewable.io -->
